presto-hive/src/main/java/io/prestosql/plugin/hive/util/FieldSetterFactory.java

```diff
@@ -365,22 +373,51 @@ public void setField(Block block, int position)
             extends FieldSetter
     {
         private final DateTimeZone timeZone;
+        private final TimestampType type;
         private final TimestampWritableV2 value = new TimestampWritableV2();
 
-        public TimestampFieldSetter(SettableStructObjectInspector rowInspector, Object row, StructField field, DateTimeZone timeZone)
+        public TimestampFieldSetter(SettableStructObjectInspector rowInspector, Object row, StructField field, TimestampType type, DateTimeZone timeZone)
         {
             super(rowInspector, row, field);
+            this.type = requireNonNull(type, "type is null");
             this.timeZone = requireNonNull(timeZone, "timeZone is null");
         }
 
         @Override
         public void setField(Block block, int position)
         {
-            long epochMilli = floorDiv(TIMESTAMP_MILLIS.getLong(block, position), MICROSECONDS_PER_MILLISECOND);
-            epochMilli = timeZone.convertLocalToUTC(epochMilli, false);
-            value.set(Timestamp.ofEpochMilli(epochMilli));
+            long epochMicros;
+            long picosOfSecond;
+            if (type.isShort()) {
+                epochMicros = type.getLong(block, position);
+                picosOfSecond = (long) floorMod(epochMicros, MICROSECONDS_PER_SECOND) * PICOSECONDS_PER_MICROSECOND;
+            }
+            else {
+                LongTimestamp longTimestamp = (LongTimestamp) type.getObject(block, position);
+                epochMicros = longTimestamp.getEpochMicros();
+                picosOfSecond = (long) floorMod(epochMicros, MICROSECONDS_PER_SECOND) * PICOSECONDS_PER_MICROSECOND +
+                        longTimestamp.getPicosOfMicro();
+            }
+            long epochSeconds = floorDiv(epochMicros, MICROSECONDS_PER_SECOND);
+            epochSeconds = convertLocalEpochSecondsToUtc(epochSeconds);
+            int nanosOfSecond = toIntExact(roundDiv(picosOfSecond, PICOSECONDS_PER_NANOSECOND));
+            if (nanosOfSecond == NANOSECONDS_PER_SECOND) {
+                epochSeconds++;
+                nanosOfSecond = 0;
+            }
+
+            Timestamp timestamp = Timestamp.ofEpochSecond(epochSeconds, nanosOfSecond);
+            value.set(timestamp);
             rowInspector.setStructFieldData(row, field, value);
         }
+
+        private long convertLocalEpochSecondsToUtc(long epochSeconds)
+        {
+            long epochMillis = epochSeconds * MILLISECONDS_PER_SECOND;
+            epochMillis = timeZone.convertLocalToUTC(epochMillis, false);
+            verify(epochMillis % MILLISECONDS_PER_SECOND == 0, "Invalid epochMillis after converting epochSeconds %s with zone %s: %s", epochSeconds, timeZone, epochMillis);
```

Comment:
We don't need this check as time zones have second precision at most. This would be a major bug in Joda-Time that we shouldn't need to verify for every call.

Reply:
@electrum see https://github.com/prestosql/presto/pull/5283#discussion_r495842141